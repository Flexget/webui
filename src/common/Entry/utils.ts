import { RawEntry, CardType, FieldNames, MappingType, GettersType } from './types';
import { Entry, DefaultEntry } from './fields';
import {
  RawMovieEntry,
  MovieEntry,
  movieFieldList,
  MovieFields,
  MovieFieldNames,
} from './fields/movies';
import {
  RawSeriesEntry,
  SeriesEntry,
  seriesFieldList,
  SeriesFields,
  SeriesFieldNames,
} from './fields/series';

import {
  RawEpisodeEntry,
  EpisodeEntry,
  EpisodeFields,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !!entry.seriesSeason;
const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<MovieFields>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<SeriesFields>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<EpisodeFields>(EpisodeFieldNames, episodesFieldList);

const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    type: CardType.Movie,
    ...getMovieProps(entry),
  };
};

const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  type: CardType.Series,
  ...getSeriesProps(entry),
});

const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  type: CardType.Episode,
  ...getEpisodeProps(entry),
});

const toDefaultEntry = (entry: RawEntry): DefaultEntry => ({
  ...entry,
  type: CardType.Default,
});

export const toEntry = (entry: RawEntry): Entry => {
  if (isMovie(entry)) {
    return toMovieEntry(entry);
  }
  if (isSeries(entry)) {
    return toSeriesEntry(entry);
  }
  if (isEpisode(entry)) {
    return toEpisodeEntry(entry);
  }
  return toDefaultEntry(entry);
};
