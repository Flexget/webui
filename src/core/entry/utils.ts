import { RawEntry, FieldNames, MappingType, GettersType, BaseEntry } from './types';
import { MovieEntry, movieFieldList, RawMovieEntry, MovieFieldNames } from './fields/movies';
import { SeriesEntry, seriesFieldList, SeriesFieldNames, RawSeriesEntry } from './fields/series';
import {
  EpisodeEntry,
  RawEpisodeEntry,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

export const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
export const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !entry.seriesSeason;
export const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry & SeriesEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode && entry.seriesName);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<RawMovieEntry>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<RawSeriesEntry>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<RawEpisodeEntry>(EpisodeFieldNames, episodesFieldList);

export const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    ...getMovieProps(entry),
  };
};

export const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  ...getSeriesProps(entry),
});

export const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  ...getEpisodeProps(entry),
});

export const toDefaultEntry = (entry: BaseEntry): BaseEntry => ({
  ...entry,
});
