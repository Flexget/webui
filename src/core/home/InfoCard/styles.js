import styled from '@emotion/styled';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardHeader from '@material-ui/core/CardHeader';
import theme from 'theme';

export const InfoCardWrapper = styled(Card)`
  margin: 0 auto;
  ${theme.breakpoints.up('sm')} {
    width: 50%;
  }
`;

export const InfoCardHeader = styled(CardHeader)`
  background-color: ${theme.palette.primary[800]};
`;

export const BoldParagraph = styled.p`
  font-weight: bold;
`;

export const InfoCardActions = styled(CardActions)`
  justify-content: center;
`;
