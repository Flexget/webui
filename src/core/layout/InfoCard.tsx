import React, { FC } from 'react';
import { css } from '@emotion/core';
import {
  IconButton,
  CardActions,
  Theme,
  Link,
  Tooltip,
  Typography,
  Dialog,
  DialogTitle,
  DialogContent,
} from '@material-ui/core';
import { GitHub, ForumOutlined, Home, Chat } from '@material-ui/icons';
import YAML from 'yaml';

const bold = css`
  font-weight: bold;
`;

const wrapper = (theme: Theme) => css`
  margin: 0 auto;
  ${theme.breakpoints.up('sm')} {
    width: 50%;
  }
`;

const actions = css`
  justify-content: center;
`;

interface Props {
  open: boolean;
  onClose: () => void;
}

const InfoCard: FC<Props> = ({ open, onClose }) => (
  <Dialog css={wrapper} open={open} onClose={onClose}>
    <DialogTitle>Flexget Web Interface</DialogTitle>
    <DialogContent dividers>
      <Typography css={bold} gutterBottom>
        We need your help! If you are a React developer or can help with the layout/design/css then
        please join in the effort!
      </Typography>
      <Typography gutterBottom>
        This is the new webui written from the ground up in React and Typescript. Please report back
        to us on how well it works, issues, ideas etc...
      </Typography>
      <Typography gutterBottom>
        There is a functional API with documentation available at <Link href="/api">/api</Link>
      </Typography>
      <Typography>
        The new webui has feature parity with v1. But if for whatever reason, if v2 is not suiting{' '}
        your needs you can add the following to your config and then you can visit{' '}
        <Link href="/v1">/v1</Link> to access old webui.
      </Typography>
      <pre>
        {YAML.stringify({
          // eslint-disable-next-line @typescript-eslint/camelcase
          web_server: {
            // eslint-disable-next-line @typescript-eslint/camelcase
            run_v1: true,
          },
        })}
      </pre>
      <Typography gutterBottom>
        More information:{' '}
        <Link href="http://flexget.com/Web-UI/v2" target="_blank" rel="noopener noreferrer">
          http://flexget.com/Web-UI/v2
        </Link>
      </Typography>
      <Typography gutterBottom>
        Chat:{' '}
        <Link href="https://flexget.com/Chat" target="_blank" rel="noopener noreferrer">
          https://flexget.com/Chat
        </Link>
      </Typography>
    </DialogContent>
    <CardActions css={actions}>
      <Tooltip title="Github">
        <IconButton
          aria-label="Github"
          href="https://github.com/Flexget/Flexget"
          target="_blank"
          rel="noopener noreferrer"
        >
          <GitHub />
        </IconButton>
      </Tooltip>
      <Tooltip title="Flexget.com">
        <IconButton
          aria-label="Flexget.com"
          href="https://flexget.com"
          target="_blank"
          rel="noopener noreferrer"
        >
          <Home />
        </IconButton>
      </Tooltip>
      <Tooltip title="Chat">
        <IconButton
          aria-label="Chat"
          href="https://flexget.com/Chat"
          target="_blank"
          rel="noopener noreferrer"
        >
          <Chat />
        </IconButton>
      </Tooltip>
      <Tooltip title="Forum">
        <IconButton
          aria-label="Forum"
          href="https://discuss.flexget.com"
          target="_blank"
          rel="noopener noreferrer"
        >
          <ForumOutlined />
        </IconButton>
      </Tooltip>
    </CardActions>
  </Dialog>
);

export default InfoCard;
