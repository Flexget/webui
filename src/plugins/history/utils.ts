import { groupBy } from 'utils/array';
import { HistoryItem, GroupByFields } from './types';

export const getGroupedHistory = (items: HistoryItem[], grouping: GroupByFields) =>
  groupBy(
    items,
    grouping,
    grouping === GroupByFields.Time ? (i: string) => i.split('T')[0] : undefined,
  );
