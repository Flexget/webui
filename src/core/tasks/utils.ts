import { RawEntry } from 'core/entry/types';

export const toExecuteRequest = (entry?: RawEntry) => {
  if (entry) {
    const { id, title, originalUrl, ...fields } = entry;
    return { title, url: originalUrl, fields };
  }
  return { title: '', url: '', fields: {} };
};
