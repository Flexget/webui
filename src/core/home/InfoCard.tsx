import React from 'react';
import { css } from '@emotion/core';
import { Card, CardContent, IconButton, CardHeader, CardActions } from '@material-ui/core';
import GitHubIcon from '@material-ui/icons/GitHub';
import ForumIcon from '@material-ui/icons/ForumOutlined';
import HomeIcon from '@material-ui/icons/Home';
import ChatIcon from '@material-ui/icons/Chat';
import theme from 'core/theme';
import { backgroundColor } from 'common/styles';

const bold = css`
  font-weight: bold;
`;

const wrapper = css`
  margin: 0 auto;
  ${theme.breakpoints.up('sm')} {
    width: 50%;
  }
`;

const actions = css`
  justify-content: center;
`;

const InfoCard = () => (
  <Card css={wrapper}>
    <CardHeader
      title="Flexget Web Interface"
      subheader="Under Development"
      css={backgroundColor(theme.palette.primary[800])}
    />
    <CardContent>
      <p css={bold}>
        We need your help! If you are a React developer or can help with the layout/design/css then
        please join in the effort!
      </p>
      <p>The interface is not yet ready for end users. Consider this preview only state.</p>
      <p>
        If you still use it anyways, please report back to us on how well it works, issues, ideas
        etc...
      </p>
      <p>
        There is a functional API with documentation available at <a href="/api">/api</a>
      </p>
      <p>
        More information:{' '}
        <a href="http://flexget.com/Web-UI/v2" target="_blank" rel="noopener noreferrer">
          http://flexget.com/Web-UI/v2
        </a>
      </p>
      <p>
        Chat:{' '}
        <a href="https://flexget.com/Chat" target="_blank" rel="noopener noreferrer">
          https://flexget.com/Chat
        </a>
      </p>
    </CardContent>
    <CardActions css={actions}>
      <IconButton
        aria-label="Github"
        href="https://github.com/Flexget/Flexget"
        target="_blank"
        rel="noopener noreferrer"
      >
        <GitHubIcon />
      </IconButton>
      <IconButton
        aria-label="Flexget.com"
        href="https://flexget.com"
        target="_blank"
        rel="noopener noreferrer"
      >
        <HomeIcon />
      </IconButton>
      <IconButton
        aria-label="Chat"
        href="https://flexget.com/Chat"
        target="_blank"
        rel="noopener noreferrer"
      >
        <ChatIcon />
      </IconButton>
      <IconButton
        aria-label="Forum"
        href="https://discuss.flexget.com"
        target="_blank"
        rel="noopener noreferrer"
      >
        <ForumIcon />
      </IconButton>
    </CardActions>
  </Card>
);

export default InfoCard;
