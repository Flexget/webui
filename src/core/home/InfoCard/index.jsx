import React from 'react';
import CardContent from '@material-ui/core/CardContent';
import IconButton from '@material-ui/core/IconButton';
import GitHubIcon from '@material-ui/icons/GitHub';
import ForumIcon from '@material-ui/icons/ForumOutlined';
import HomeIcon from '@material-ui/icons/Home';
import ChatIcon from '@material-ui/icons/Chat';
import { InfoCardWrapper, InfoCardHeader, BoldParagraph, InfoCardActions } from './styles';

const InfoCard = () => (
  <InfoCardWrapper>
    <InfoCardHeader title="Flexget Web Interface" subheader="Under Development" />
    <CardContent>
      <BoldParagraph>
        We need your help! If you are a React developer or can help with the layout/design/css then
        please join in the effort!
      </BoldParagraph>
      <p>The interface is not yet ready for end users. Consider this preview only state.</p>
      <p>
        If you still use it anyways, please report back to us on how well it works, issues, ideas
        etc...
      </p>
      <p>
        There is a functional API with documentation available at <a href="/api">/api</a>
      </p>
      <p>
        More information:{' '}
        <a href="http://flexget.com/Web-UI/v2" target="_blank" rel="noopener noreferrer">
          http://flexget.com/Web-UI/v2
        </a>
      </p>
      <p>
        Chat:{' '}
        <a href="https://flexget.com/Chat" target="_blank" rel="noopener noreferrer">
          https://flexget.com/Chat
        </a>
      </p>
    </CardContent>
    <InfoCardActions>
      <IconButton
        aria-label="Github"
        href="https://github.com/Flexget/Flexget"
        target="_blank"
        rel="noopener noreferrer"
      >
        <GitHubIcon />
      </IconButton>
      <IconButton
        aria-label="Flexget.com"
        href="https://flexget.com"
        target="_blank"
        rel="noopener noreferrer"
      >
        <HomeIcon />
      </IconButton>
      <IconButton
        aria-label="Chat"
        href="https://flexget.com/Chat"
        target="_blank"
        rel="noopener noreferrer"
      >
        <ChatIcon />
      </IconButton>
      <IconButton
        aria-label="Forum"
        href="https://discuss.flexget.com"
        target="_blank"
        rel="noopener noreferrer"
      >
        <ForumIcon />
      </IconButton>
    </InfoCardActions>
  </InfoCardWrapper>
);

export default InfoCard;
