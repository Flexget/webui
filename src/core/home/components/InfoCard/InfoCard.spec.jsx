import React from 'react';
import renderer from 'react-test-renderer';
import InfoCard from 'core/home/components/InfoCard';
import { themed } from 'utils/tests';

describe('core/home/components/InfoCard', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      themed(<InfoCard />),
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
