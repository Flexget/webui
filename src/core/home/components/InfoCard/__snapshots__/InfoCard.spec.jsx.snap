// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/home/components/InfoCard renders correctly 1`] = `
.glamor-3 {
  margin: 0 auto;
}

.glamor-0 {
  background-color: #ef6c00;
}

.glamor-1 {
  font-weight: bold;
}

.glamor-2 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (min-width:600px) {
  .glamor-3 {
    width: 50%;
  }
}

<div
  className="MuiPaper-root-1 MuiPaper-shadow2-5 MuiPaper-rounded-2 glamor-3 e17d04jh0"
>
  <div
    className="MuiCardHeader-root-28 glamor-0 e17d04jh1"
  >
    <div
      className="MuiCardHeader-content-31"
    >
      <span
        className="MuiTypography-root-34 MuiTypography-headline-39 MuiCardHeader-title-32"
      >
        Flexget Web Interface
      </span>
      <span
        className="MuiTypography-root-34 MuiTypography-body1-43 MuiTypography-colorTextSecondary-56 MuiCardHeader-subheader-33"
      >
        Under Development
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root-58"
  >
    <p
      className="glamor-1 e17d04jh2"
    >
      We need your help! If you are a React developer or can help with the layout/design/css then please join in the effort!
    </p>
    <p>
      The interface is not yet ready for end users. Consider this preview only state.
    </p>
    <p>
      If you still use it anyways, please report back to us on how well it works, issues, ideas etc...
    </p>
    <p>
      There is a functional API with documentation available at
       
      <a
        href="/api"
      >
        /api
      </a>
    </p>
    <p>
      More information:
      <a
        href="http://flexget.com/wiki/Web-UI/v2"
        rel="noopener noreferrer"
        target="_blank"
      >
        http://flexget.com/wiki/Web-UI/v2
      </a>
    </p>
    <p>
      Gitter Chat:
      <a
        href="https://gitter.im/Flexget/Flexget"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://gitter.im/Flexget/Flexget
      </a>
    </p>
  </div>
  <div
    className="MuiCardActions-root-59 glamor-2 e17d04jh3"
  >
    <a
      aria-label="Github"
      className="MuiButtonBase-root-67 MuiIconButton-root-61 MuiCardActions-action-60"
      href="https://github.com/Flexget/Flexget"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiIconButton-label-66"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-69 MuiIcon-fontSize-75 fa fa-github"
        />
      </span>
      <span
        className="MuiTouchRipple-root-76"
      />
    </a>
    <a
      aria-label="Flexget.com"
      className="MuiButtonBase-root-67 MuiIconButton-root-61 MuiCardActions-action-60"
      href="https://flexget.com"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiIconButton-label-66"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-69 MuiIcon-fontSize-75 fa fa-home"
        />
      </span>
      <span
        className="MuiTouchRipple-root-76"
      />
    </a>
    <a
      aria-label="Gitter"
      className="MuiButtonBase-root-67 MuiIconButton-root-61 MuiCardActions-action-60"
      href="https://gitter.im/Flexget/Flexget"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiIconButton-label-66"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-69 MuiIcon-fontSize-75 fa fa-comment"
        />
      </span>
      <span
        className="MuiTouchRipple-root-76"
      />
    </a>
    <a
      aria-label="Forum"
      className="MuiButtonBase-root-67 MuiIconButton-root-61 MuiCardActions-action-60"
      href="https://discuss.flexget.com"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiIconButton-label-66"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-69 MuiIcon-fontSize-75 fa fa-forumbee"
        />
      </span>
      <span
        className="MuiTouchRipple-root-76"
      />
    </a>
  </div>
</div>
`;
